File: Offer/2.py
# 使用 list 实现
class Solution:
    def replaceSpace(self, s):
        if s == '' or not isinstance(s, str):
            return ''
        s_list = list(s)
        new_str = []
        for item in s_list:
            if item == ' ':
                new_str.extend('%20')
            else:
                new_str.append(item)
        return ''.join(new_str)


# 使用从后向前的遍历
class Solution:
    def replaceSpace(self, s):
        if s == '' or not isinstance(s, str):
            return ''
        new_len = 0
        for item in s:
            if item == ' ':
                new_len += 2
            new_len += 1
        new_list = new_len * [None]
        j = new_len - 1
        for i in range(len(s) - 1, -1, -1):
            if s[i] == ' ':
                new_list[j] = '0'
                new_list[j - 1] = '2'
                new_list[j - 2] = '%'
                j -= 3
                i -= 1
            else:
                new_list[j] = s[i]
                j -= 1
                i -= 1
        return ''.join(new_list)

File: Offer/offerCPP/2.cpp
// in-place 操作
// 从后往前 O(n)
class Solution
{
  public:
    void replaceSpace(char *str, int length)
    {
        // 判断特殊情况
        if (str == nullptr || length <= 0)
            return;
        int newlen = 0;
        int oldlen = 0;
        // 统计原字符串和新字符串的长度
        int j = 0;  // 从头开始
        while (str[j] != '\0')
        {
            newlen++;
            oldlen++;
            if (str[j] == ' ')
                newlen += 2;
            j++;
        }
        // 新长度大于总长度
        if (newlen > length || newlen == oldlen)
            return;
        str[newlen] = '\0';  // 以 '\0' 结束字符串
        // 两个指针
        for (int i = oldlen - 1, j = newlen - 1; i >= 0; i--, j--)
        {
            if (str[i] == ' ')
            {
                str[j] = '0';
                str[j - 1] = '2';
                str[j - 2] = '%';
                j = j - 2;
            }
            else
                str[j] = str[i];  // 复制移动
        }
    }
};
